rosidl_typesupport_opendds_c/resource/srv__type_support_c.cpp.em

```diff
@@ -132,71 +180,187 @@ static void * create_replier__@(service.namespaced_type.name)(
     DDS::Subscriber_var dds_subscriber,
     allocator_t allocator)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
-  return NULL;
+  return @('::'.join(service.namespaced_type.namespaces))::typesupport_opendds_cpp::create_replier__@(service.namespaced_type.name)(
+    dds_participant,
+    request_topic_str,
+    response_topic_str,
+    dds_publisher,
+    dds_subscriber,
+    allocator);
 }
 
 static const char * destroy_replier__@(service.namespaced_type.name)(
   void * untyped_replier,
   deallocator_t deallocator)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
-  return NULL;
+  return @('::'.join(service.namespaced_type.namespaces))::typesupport_opendds_cpp::destroy_replier__@(service.namespaced_type.name)(
+    untyped_replier, deallocator);
 }
 
 static bool take_request__@(service.namespaced_type.name)(
   void * untyped_replier,
   rmw_request_id_t * request_header,
   void * untyped_ros_request)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
-  return true;
+  using ReplierType = rosidl_typesupport_opendds_cpp::Replier<
+    @(__dds_msg_typesupport_type)RequestWrapper,
+    @(__dds_msg_typesupport_type)ResponseWrapper
+  >;
+  if (!untyped_replier || !request_header || !untyped_ros_request) {
+    return false;
+  }
+  
+  ReplierType * replier = static_cast<ReplierType *>(untyped_replier);
+
+  @(__dds_msg_typesupport_type)RequestWrapper dds_request_wrapper;
+
+  if (DDS::RETCODE_OK != replier->take_request(dds_request_wrapper)) {
+    fprintf(stderr, "take_request failed\n");
+    return false;
+  }
+
+  int64_t sequence_number =
+    (((int64_t)dds_request_wrapper.header().request_id().sequence_number().high) << 32) |
+    dds_request_wrapper.header().request_id().sequence_number().low;
+  request_header->sequence_number = sequence_number;
+
+  size_t IDENTITY_SIZE = 16;
+  OpenDDS::DCPS::GUID_t id = dds_request_wrapper.header().request_id().writer_guid();
+  std::memcpy(&(request_header->writer_guid[0]), &id, IDENTITY_SIZE);
+
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(
+    rosidl_typesupport_opendds_c,
+    @(', '.join(service.request_message.structure.namespaced_type.namespaces)),
+    @(service.request_message.structure.namespaced_type.name))();
+  const message_type_support_callbacks_t * callbacks =
+    static_cast<const message_type_support_callbacks_t *>(ts->data);
+  bool converted = callbacks->convert_dds_to_ros(&(dds_request_wrapper.request()), untyped_ros_request);
```

Comment:
Combine with line 240?  That is, just make this the return expression.

Reply:
Fixed